src/app/fdctl/run/tiles/fd_repair.c

```diff
@@ -0,0 +1,675 @@
+      FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz,
+            ctx->contact_in_chunk0, ctx->contact_in_wmark ));
+
+    uchar const * dcache_entry = fd_chunk_to_laddr_const( ctx->contact_in_mem, chunk );
+    handle_new_cluster_contact_info( ctx, dcache_entry );
+    return;
+  }
+
+  if( FD_UNLIKELY( in_idx==STAKE_IN_IDX ) ) {
+    if( FD_UNLIKELY( chunk<ctx->stake_weights_in_chunk0 || chunk>ctx->stake_weights_in_wmark ) )
+      FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz,
+            ctx->stake_weights_in_chunk0, ctx->stake_weights_in_wmark ));
+
+    uchar const * dcache_entry = fd_chunk_to_laddr_const( ctx->stake_weights_in_mem, chunk );
+    handle_new_stake_weights( ctx, dcache_entry );
+    return;
+  }
+
+  if( FD_UNLIKELY( in_idx==STORE_IN_IDX ) ) {
+    if( FD_UNLIKELY( chunk<ctx->repair_req_in_chunk0 || chunk>ctx->repair_req_in_wmark ) )
+      FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz,
+            ctx->repair_req_in_chunk0, ctx->repair_req_in_wmark ));
+
+    uchar const * dcache_entry = fd_chunk_to_laddr_const( ctx->repair_req_in_mem, chunk );
+    handle_new_repair_requests( ctx, dcache_entry, sz );
+    return;
+  }
+  
+  if( FD_UNLIKELY( chunk<ctx->chunk || chunk>ctx->wmark || sz>FD_NET_MTU ) ) {
+    FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz, ctx->chunk, ctx->wmark ));
+    *opt_filter = 1;
+    return;
+  }
+
+  uchar const * dcache_entry = fd_chunk_to_laddr_const( ctx->net_in, chunk );
+  ulong  hdr_sz = fd_disco_netmux_sig_hdr_sz( sig );
+  FD_TEST( hdr_sz < sz ); /* Should be ensured by the net tile */
+
+  ctx->repair_buffer_sz = sz;
+  fd_memcpy( ctx->repair_buffer, dcache_entry, sz );
+
+  *opt_filter = 0;
+
+  return;
+}
+
+static void
+after_frag( void *             _ctx,
+            ulong              in_idx,
+            ulong              seq        FD_PARAM_UNUSED,
+            ulong *            opt_sig,
+            ulong *            opt_chunk  FD_PARAM_UNUSED,
+            ulong *            opt_sz     FD_PARAM_UNUSED,
+            ulong *            opt_tsorig FD_PARAM_UNUSED,
+            int *              opt_filter,
+            fd_mux_context_t * mux        FD_PARAM_UNUSED) {
+
+  fd_repair_tile_ctx_t * ctx = (fd_repair_tile_ctx_t *)_ctx;
+
+  *opt_filter = 1;
+
+  if( FD_UNLIKELY( in_idx==CONTACT_IN_IDX ) ) {
+    return;
+  }
+
+  if( FD_UNLIKELY( in_idx==STAKE_IN_IDX ) ) {
+    return;
+  }
+
+  if( FD_UNLIKELY( in_idx==STORE_IN_IDX ) ) {
+    return;
+  }
+
+  uint   ip = fd_disco_netmux_sig_dst_ip( *opt_sig );
+  ulong hdr_sz = fd_disco_netmux_sig_hdr_sz( *opt_sig );
+  eth_ip_udp_t * hdr = (eth_ip_udp_t *)ctx->repair_buffer;
+
+  fd_repair_peer_addr_t peer_addr;
+  peer_addr.l = 0;
+  peer_addr.addr = ip;
+  peer_addr.port = fd_ushort_bswap( hdr->udp->net_sport );
+
+  fd_repair_settime( ctx->repair, fd_log_wallclock() );
+  fd_repair_continue( ctx->repair );
```

Comment:
Does this turn the internal engine of the repair module? You probably don't want to tie this to incoming frags since you might never get any, but still want to turn the engine. Continue should probably go in after_credit?

Reply:
continue primarily sends repair requests and prints stats. i don't think this instance needs to shift

Reply:
the continue is to trigger delayed actions which are time dependent, such as shuffling peers or sending pushes